### PR TITLE
Instant Search: Improve product result format layout

### PR DESCRIPTION
**Pull request by jsnmoon**
#### Changes proposed in this Pull Request:
* Adds a 5-column layout to the product result format at large breakpoints.

Screen Shot 2021-03-24 at 5 48 48 PM

* Adds a placeholder visual if the result lacks a corresponding image.

Screen Shot 2021-03-24 at 5 49 00 PM

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
* Apply this change to your site with a Jetpack Search subscription.
* Perform a site search; append `result_format=product` query string to force the product result format layout. Ensure that the results look reasonable on all breakpoints (e.g. desktop, tablet, mobile).

### CODE REVIEW

**Comment by bluefuton** [comment]
The 4-per-row works well on bigger screens 👍 I like it!

It would be good if the placeholder image could match the height of the row, so the product titles lined up. It's like that in your example screenshot, but on my test site:

Screen Shot 2021-03-05 at 14 11 35

**Comment by jsnmoon** [comment]
> Screen Shot 2021-02-17 at 4 07 11 PM

> Screen Shot 2021-03-05 at 14 11 35

@jeffgolenski: Could you recommend a better placeholder design than the one already implemented above? The landscape ratio of the image gridicon doesn't seem to lend itself to the grid layout.

**Review by bluefuton — DISMISSED** [review]
I think the image placeholder works well enough for the moment :) Just two small things to fix:

- Line up titles for products with images and products with placeholders

Screen Shot 2021-03-10 at 16 15 25

- Update the SVG title for the placeholder (isn't correct in this case):

Screen Shot 2021-03-10 at 16 13 17

**Review by bluefuton — COMMENTED** [review]
I think you mentioned this earlier today - getting some inconsistent image heights (Photon is turned off):

Screen Shot 2021-03-12 at 16 14 09

Titles are now lining up when images are big enough, but perhaps need a tiny bit more top margin on the title? (cc @jeffgolenski)

IE11 has some row height issues:

Screen Shot 2021-03-12 at 16 14 44

**Comment by jsnmoon** [comment]
@bluefuton: I went ahead and changed to a static layout for the product grid. This means that images won't automatically resize to fit the grid, but it guarantees height/width for each image.

**Comment by bluefuton** [comment]
One possible solution for IE object-fit: https://www.stevefenton.co.uk/2019/09/fixing-css-object-fit-for-internet-explorer/

It'd be nice if we could use more of the available horizontal space (particularly on bigger screens) but this is much more solid across browsers 👍

Screen Shot 2021-03-15 at 14 34 37

**Comment by jsnmoon** [comment]
Further reduced the grid width to increase content density. I'll punt the object-fit fix for IE11 to our image component enhancement PR.

**Comment by bluefuton** [comment]
This is how it currently displays on my usual browser window size:

Screen Shot 2021-03-17 at 16 04 32

Is there an opportunity to use a more flexible grid? (Like https://css-tricks.com/designing-a-product-page-layout-with-flexbox/ ?) Are we constrained by image size?

**Comment by jsnmoon** [comment]
I was able to restore a dynamic flexbox-based layout using the [`aspect-ratio` style rule](https://caniuse.com/mdn-css_properties_aspect-ratio), which appears to be decently supported by most browsers.

It doesn't work with IE11, though.

**Comment by bluefuton** [comment]
Thanks @jsnmoon! Is it something other than `aspect-ratio` causing the wonky line height in IE11? I see `aspect-ratio` isn't well supported in current Safari or Firefox either but the product rows look consistent there.

Screen Shot 2021-03-19 at 12 02 47

**Comment by jsnmoon** [comment]
> Is it something other than `aspect-ratio` causing the wonky line height in IE11? I see `aspect-ratio` isn't well supported in current Safari or Firefox either but the product rows look consistent there.

I decided against using `aspect-ratio` due to its poor support. In its place, I used [a CSS hack to implement a constant aspect ratio box](https://css-tricks.com/aspect-ratio-boxes/).

This solution appears to work reliably for Chrome, Firefox, Safari, and IE11.

**Comment by jsnmoon** [comment]
Updated placeholder visuals when a search result contains no images:

Screen Shot 2021-03-19 at 3 30 46 PM

**Comment by bluefuton** [comment]
> Updated placeholder visuals when a search result contains no images 

Shall we clamp the description/body text to say, 2 lines for product layout? cc @jeffgolenski 

Screen Shot 2021-03-22 at 15 26 44

**Comment by bluefuton** [comment]
Layout is looking nice across browsers now 🎉

One thing I noticed - the sort selector is unresponsive when positioned on the right. Perhaps hidden behind another element?

Screen Shot 2021-03-22 at 15 49 00

**Comment by jsnmoon** [comment]
> Shall we clamp the description/body text to say, 2 lines for product layout? cc @jeffgolenski

I went ahead and replaced the highlights with a keyword match snippet from the design specifications:

Screen Shot 2021-03-22 at 2 02 41 PM

> One thing I noticed - the sort selector is unresponsive when positioned on the right. Perhaps hidden behind another element?

This was a z-index issue and has now been fixed.

**Review by bluefuton — COMMENTED** [review]
Sort now works again. Thanks! 🎉 

Good call on "keyword found in content" @jsnmoon 👍 It's appearing in bold for me, which draws the eye a bit too much - can we set the font weight to normal?

Screen Shot 2021-03-23 at 11 53 36

One other bit of feedback - it says "keyword found in content" even when I haven't entered a search string, which feels a bit weird. Can we omit it when the search string is empty?

**Comment by bluefuton** [comment]
One more request as I'm looking at adding `alt` text for images: is it straightforward to use a `img` element rather than a background-image?

Screen Shot 2021-03-23 at 12 34 38

**Comment by jsnmoon** [comment]
> Good call on "keyword found in content" @jsnmoon 👍 It's appearing in bold for me, which draws the eye a bit too much - can we set the font weight to normal? (@bluefuton)

Good catch, this is likely caused by the theme. 

> One other bit of feedback - it says "keyword found in content" even when I haven't entered a search string, which feels a bit weird. Can we omit it when the search string is empty? (@bluefuton)

I'm slightly against this idea since our API is essentially indicating a keyword match via the result's `highlight` field. I'll go ahead and make the change here for now, but let's prioritize improving the API's `highlight` output in the future (e.g. returning `null` instead of `[""]` for non-matches).

> One more request as I'm looking at adding `alt` text for images: is it straightforward to use a `img` element rather than a background-image? (@bluefuton)

Done! I went ahead and set the `alt` to the result's `title` for now.

File: projects/plugins/jetpack/modules/search/instant-search/components/search-result-product.scss
Comment (jsnmoon) [inline projects/plugins/jetpack/modules/search/instant-search/components/search-result-product.scss:67]: These styles are no longer needed, will delete soon.

**Review by bluefuton — COMMENTED** [review]
Looking so much more polished now! Excited to get this launched :)

I found one display issue with long, tall portrait images obscuring the title (do we need a `max-height`?):

Screen Shot 2021-03-24 at 12 23 41

This is the errant image:

1000071377_070_Zoom_3-Full_Length

**Comment by bluefuton** [comment]
> I'm slightly against this idea since our API is essentially indicating a keyword match via the result's highlight field. I'll go ahead and make the change here for now, but let's prioritize improving the API's highlight output in the future

Created https://github.com/Automattic/jetpack/issues/19272 to keep track of this 👍

**Comment by jsnmoon** [comment]
> I found one display issue with long, tall portrait images obscuring the title (do we need a max-height?):

Looks like this was a regression caused by switching from `div` to `img`. I've fixed it using `height`, `width`, and `object-fit` styles.

**Comment by jeffgolenski** [comment]
Hey hey @jsnmoon & @bluefuton this is really starting to look AMAZING. 

I'm running this branch live on midnightshift.photo because I'm reckless. haha

Here's a quick design review so far with a video of various screen sizes. Please ignore anything you're aware of:


https://user-images.githubusercontent.com/214813/112383799-f4e39900-8cc3-11eb-920e-331465a40338.mp4

* Image scaling is strange with the gray block behind the image
* Results that highlight the keyword in the title, don't need to have the "keyword found in content" label
* The sorting is functional and works great—we should just clean up the font and sizing of the box there. Let's make the drop down UI a bit bigger and the font a bit larger
* Can we make "clear" a little larger as well?
* On mobile / small screens, can the sort by occupy the entire width of the screen for easier tapping

* ~0:55 in video: When a search phrase is removed and the field is empty are we just displaying whatever is found in /shop/? In this instance, I like how you've defaulted it to say "showing popular results." Do you think there's a need to remind people to enter a phrase when that field is empty?

**Comment by jsnmoon** [comment]
> - [x] Image scaling is strange with the gray block behind the image (@jeffgolenski)

Fixed in df604b0e6ab58982609608e03b3ce6165dc9eea4!

> - [x] Results that highlight the keyword in the title, don't need to have the "keyword found in content" label (@jeffgolenski)

Fixed in c1fec7950c4c681720081ea930f729cc14e6c699!

> - [x] The sorting is functional and works great—we should just clean up the font and sizing of the box there. Let's make the drop down UI a bit bigger and the font a bit larger (@jeffgolenski)
> - [x] Can we make "clear" a little larger as well? (@jeffgolenski)
> - [x] On mobile / small screens, can the sort by occupy the entire width of the screen for easier tapping (@jeffgolenski)

Done!

> - [x] When a search phrase is removed and the field is empty are we just displaying whatever is found in /shop/? (@jeffgolenski)

I think so, yes. I'm under the impression that we might be scoring (and thus sorting) the default popular results differently than in `/shop/`.

> Do you think there's a need to remind people to enter a phrase when that field is empty? (@jeffgolenski)

I don't think that's necessary 😊

---

I also made two additional changes:
- Reduced the verbosity of the match hint to say `Matches content`/`Matches comment` instead of `Keyword found in content`/`Keyword found in comment`.
- Reduced the verbosity of the sort label to say `Sort:` instead of `Sort by:`.

**Comment by jsnmoon** [comment]
I refactored `SearchControls` out of the `SearchForm` component to simplify component styling. This can technically impact other result formats, so please test minimal and expanded result formats if you're able.

**Review by bluefuton — APPROVED** [review]
Lovely. Tested out well for me on Firefox, Chrome and Edge.

Two minor nits but these don't need to hold up the approval:

*  The "Showing popular results" header now feels a bit close to the sort selector on the smallest breakpoint.

Screen Shot 2021-03-25 at 15 29 47

* A search string that just contains whitespace (" ") will trigger the "matches content" message - should we trim the string before doing that check? e.g. http://example.com/?s=%20&result_format=product. Totally acknowledge that this best fixed in the API ultimately (filed in https://github.com/Automattic/jetpack/issues/19272).

**Comment by kangzj** [comment]
Awesome work @jsnmoon 👍 
Just had a spin, there was something which might need some attention. On wider screens like Jeff's, the image could get a little bit too big.

https://d.pr/i/dXe539

**Review by jeherve — APPROVED** [review]
It tested well for me. 🚢

**Comment by jsnmoon** [comment]
r223374-wpcom.

**Pull request merged**
